Documentation/RepoApi.md

```diff
@@ -0,0 +1,137 @@
+
+### Parameters
+
+Each action above will take a set of parameters used to pass information into the action and control its behavior.
+
+Parameters will take the form of `-p:ParameterName=ParameterValue`, just like MSBuild properties are passed to `msbuild.exe`.
+One slight variation here is the usage of `-p` instead of `/p`.  MSBuild supports both forms, but source-build chooses `-p` because
+it meets [POSIX recommendations](https://www.gnu.org/software/libc/manual/html_node/Argument-Syntax.html) for command line parameters
+and avoids any confusion with file paths on Unix machines.
+
+In order to avoid conflicts between existing MSBuild properties used by the repos and the properties passed in by source-build,
+source-build will append a prefix to every property it passes in: `DotNet`.
+
+#### Detecting build from source
+
+The first parameter used by source-build will be a value indicating whether the build is being performed in "source only" mode. This
+means repos cannot use tools/packages/references that haven't already been built by the current build.
+
+`-p:DotNetBuildFromSource=true`
+
+Some repos build more projects than is necessary to build the .NET Core stack. For example, the `dotnet/roslyn` repo builds
+the C# compiler, which is needed, but it also builds Visual Studio extensions, which aren't needed. These VS extensions
+cannot be built without using an external dependency on tooling that isn't available when building from source.
+
+#### Available tools
+
+In order for repos to build, they need tools in order to do the build (ex. MSBuild). These tools are not generally available
+on all supported operating systems. As such, many repos today have the ability to pull these tools from the internet during the
+build.
+
+However, there are a couple problems with each repo downloading its own toolset.
+
+1. Different versions can cause inconsistencies
+1. It takes a non-trivial amount of time to download the toolset, if every repo does it during the complete .NET Core build,
+it will add a lot of overhead to the build.
+
+To solve this, source-build will have a set of common tools available that it will pass into each repos build.
+
+First, a [.NET Core SDK](https://www.microsoft.com/net/download/core#/sdk) will be available for repos to be able to invoke
+MSBuild, NuGet, etc. The path to the folder containing the `dotnet` executable will be available using the
+
+`-p:DotNetToolDir`
+
+property. This can be used to call `dotnet restore`, `dotnet build`, `dotnet publish`, `dotnet msbuild`, or any other CLI
+verb on the repo's projects.
+
+Second, a common set of tools used by the `dotnet` org repos is [BuildTools](https://github.com/dotnet/buildtools). This will
+be available using the
+
+`-p:DotNetBuildToolsDir`
+
+property. This can be used to invoke any current or future tool that is part of BuildTools. It is not required that each repo
```

Comment:
Will this contain nuget packages for BuildTools or the contents of BuildTools repository?

Reply:
I think this is going to contain a published set of binaries that would generally come when you consume Buildtools.

Reply:
I'd like to see us get to a place where BuildTools is no longer a monolithic thing to consume, but rather specific "libraries" (packages) are chosen as needed.

Reply:
Obviously I think that is a good idea as well @markwilkie but I do think this is still going to be a pointer to some "Tools" folder that has all the components you restored in it.

Reply:
And for source-build we might want it to restore everything otherwise it has to figure out how to see just the parts that each of the repo's needs and aggregate them.